bats/tests/helpers/utils.bats

```diff
@@ -196,6 +208,130 @@ get_json_test_data() {
 
 ########################################################################
 
+@test 'semver a1b2.3c4.5.6d7.8.9.0' {
+    is 4.5.6
+}
+
+@test 'semver a1b2.3c4.5' {
+    is 2.3.0
+}
+
+@test 'semver a1b2c3' {
+    is 1.0.0
+}
+
+@test 'semver 1.2.3.4' {
+    is 1.2.3
+}
+
+@test 'semver 00.00.00' {
+    is 0.0.0
+}
+
+@test 'semver 000000' {
+    is 0.0.0
+}
+
+@test 'semver 0.001' {
+    is 0.1.0
+}
+
+@test 'semver 00100.00200.00300' {
+    is 100.200.300
+}
+
+@test 'semver ignores dates/times' {
+    # Can really pass quoted strings in the test description, so can't include spaces
+    run semver "1/1/70 12:00:00 version 7.8"
+    assert_success
+    assert_output 7.8.0
+}
+
+@test 'semver looks at all lines of the input' {
+    run semver $'Version1: 1.2\nVersion2: 3.4.5'
+    assert_success
+    assert_output 3.4.5
+}
+
+@test 'semver looks only at the first argument' {
+    run semver 'Version1: 1.2' 'Version2: 3.4.5'
+    assert_success
+    assert_output 1.2.0
+}
+
+@test 'semver fails when input has no number' {
+    run semver "Hello world"
+    assert_failure
+}
+
+########################################################################
+
+@test 'semver_eq 1.2.3 1.2.3' {
+    succeeds
+}
+@test 'semver_eq 1.2 1.2.0' {
+    fails
```

Comment:
That's going to be tricky; it makes sense given the implementation, but we'd probably expect this to succeed in actual usage.

(This might mean changing `semver_eq` to parse the two arguments separately and then compare…)

Reply:
It really is invalid input according to the function docs:

```
# Compare 2 regular major.minor.patch versions as returned by the semver function above
```

Maybe I should remove this test case.

We can argue that the `semver_xxx` functions should validate the arguments, but how would you report an error? Calling `fatal` is really the only option I can think off; should we do that?

Reply:
I think calling `fatal` if the arguments are unexpected would be good here; otherwise we can end up with a command outputting garbage, and accidentally passing `semver_le` or something.

Doesn't need to be a full parse; we can just `fatal` if not `grep --quiet -E '^[0-9]+\.[0-9]+\.[0-9]+$'` or something.

Reply:
I can't find any way how to abort a test from inside a function that is called via `run`, so "throwing an exception" is not possible.

Validation will have to be done by the caller via the new `semver_is_valid` function. 

Note that the return value of `semver` is always valid unless the function failed (in which case it returns nothing).